packages/api/test/e2e/doughnut.e2e.ts

```diff
@@ -0,0 +1,206 @@
+    holder: KeyringPair,
+    permissions: Record<string, Uint8Array>,
+) {
+    const d = new DoughnutMaker(
+        issuer.publicKey,
+        holder.publicKey,
+        (new Date().getTime()) * 2, // expiry set to twice the current timestamp
+        1, //notBefore
+    );
+    for (const key in permissions) {
+        d.addDomain(key, permissions[key]);
+    }
+    d.signSr25519(new Uint8Array(issuer.secretKey));
+    return d.encode();
+}
+
+
+const Dave = {
+    address: '5DAAnrj7VHTznn2AWBemMuyBwZWs6FNFjdyVXUeYum3PTXFy',
+    uri: '//Dave',
+};
+
+describe('Doughnut for CennznetExtrinsic', () => {
+    let aliceKeyPair = {
+        secretKey: hexToU8a('0x98319d4ff8a9508c4bb0cf0b5a78d760a0b2082c02775e6e82370816fedfff48925a225d97aa00682d6a59b95b18780c10d7032336e88f3442b42361f4a66011'),
+        publicKey: hexToU8a('0xd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d'),
+    };
+
+    let api: Api;
+    let alice, bob, charlie;
+    let holderKeyPair;
+    let keyring: {
+        [index: string]: KeyringPair;
+    };
+
+    beforeAll(async () => {
+        api = await initApiPromise();
+        await cryptoWaitReady();
+        const keyring = new Keyring({type: 'sr25519'});
+        alice = keyring.addFromUri('//Alice');
+        bob = keyring.addFromUri('//Bob');
+        charlie = keyring.addFromUri('//Charlie');
+        holderKeyPair = bob;
+    });
+
+    afterEach(() => {
+        jest.setTimeout(5000);
+    });
+
+    it('Delegates a GA transfer from alice to charlie when extrinsic is signed by bob', async done => {
+
+        let doughnut = makeDoughnut(
+            aliceKeyPair,
+            holderKeyPair,
+            {'cennznet': makeCennznut('generic-asset', 'transfer')},
+        );
+        const CENNZ = 16000;
+        const reciever = Dave.address;
+        const amount = 8767535;
+
+        const tx = api.tx.genericAsset.transfer(CENNZ, reciever, amount);
+        const doughnutB = new Doughnut(api.registry, doughnut);
+        const opt = {doughnut: doughnutB};
+        await tx.signAndSend(holderKeyPair, opt as any, async ({events, status}) => {
+            if (status.isFinalized) {
+                const blockHash = status.value.toHex();
+                console.log('Completed at block hash', blockHash);
+                const allEvents = await api.query.system.events.at(blockHash);
+                const transfer = allEvents.find(
+                    event => (
+                        event.event.data.method === 'Transferred' &&
+                        event.event.data.section === 'genericAsset' &&
+                        event.event.data[1].toString() === alice.address && // transferred from alice's account
+                        event.event.data[0].toString() === CENNZ.toString() && // CENNZ
+                        event.event.data[2].toString() === reciever && // Dave is reciever
+                        event.event.data[3].toString() === amount.toString()
+                    ),
```

Comment:
You can move the event check to the place before ```if (status.isFinalized)``` so you wouldn't need to find it in allEvents

Reply:
no...I can't... only after it is finalised I will get the block hash..from which I would be able to get the event corresponding to that block hash

Reply:
You don't need the block hash, you just need to check if the event is a generic asset transfer from Alice to Dave with the specified amount

Reply:
as discussed I'll raise a ticket for having events from callback stuff